rp2040-hal/src/uart.rs

```diff
@@ -1,3 +1,356 @@
+        stop_bits: StopBits::One,
+        parity: None
+    };
+}
+
+/// An UART Peripheral based on an underlying UART device.
+pub struct UARTPeripheral<S: State, D: UARTDevice> {
+    device: D,
+    _state: S,
+    config: UARTConfig,
+    effective_baudrate: Baud
+}
+
+impl<S: State, D: UARTDevice> UARTPeripheral<S, D> {
+    fn transition<To: State>(self, state: To) -> UARTPeripheral<To, D> {
+        UARTPeripheral {
+            device: self.device,
+            config: self.config,
+            effective_baudrate: self.effective_baudrate,
+            _state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> D{
+        self.device
+    }
+}
+
+impl<D: UARTDevice> UARTPeripheral<Disabled, D> {
+
+    /// Enables the provided UART device with the given configuration.
+    pub fn enable(mut device: D, config: UARTConfig, frequency: Hertz) -> UARTPeripheral<Enabled, D> {
+
+        let effective_baudrate = configure_baudrate(&mut device, &config.baudrate, &frequency);
+
+        // Enable the UART, both TX and RX
+        device.uartcr.write(|w| {
+            w.uarten().set_bit();
+            w.txe().set_bit();
+            w.rxe().set_bit();
+            w
+        });
+
+        device.uartlcr_h.write(|w| {
+            w.fen().set_bit();
+
+            set_format(w, &config.data_bits, &config.stop_bits, &config.parity);
+            w
+        });
+
+        device.uartdmacr.write(|w| {
+            w.txdmae().set_bit();
+            w.rxdmae().set_bit();
+            w
+        });
+
+        UARTPeripheral {
+            device, config, effective_baudrate, _state: Enabled
+        }
+    }
+}
+
+impl<D: UARTDevice> UARTPeripheral<Enabled, D> {
+
+    /// Disable this UART Peripheral, falling back to the Disabled state.
+    pub fn disable(self) -> UARTPeripheral<Disabled, D> {
+        self.transition(Disabled)
+    }
+
+    fn uart_is_writable(&self) -> bool {
+        self.device.uartfr.read().txff().bit_is_clear()
+    }
+
+    fn uart_is_readable(&self) -> bool {
+        self.device.uartfr.read().rxfe().bit_is_clear()
+    }
+
+    /// Writes bytes to the UART.
+    /// This function writes as long as it can. As soon that the FIFO is full, if :
+    /// - 0 bytes were written, a WouldBlock Error is returned
+    /// - some bytes were written, it is deemed to be a success
+    /// Upon success, the number of written bytes is returned.
+    pub fn write(&self, data: &[u8]) -> nb::Result<usize, Infallible> {
+
+        let mut bytes_written = 0;
+
+        for c in data {
+
+            if !self.uart_is_writable() {
+                if bytes_written == 0 {
+                    return Err(WouldBlock)
+                }
+                else {
+                    return Ok(bytes_written)
+                }
+            }
+
+            bytes_written += 1;
+
+            self.device.uartdr.write(|w| unsafe {
+                w.data().bits(*c);
+                w
+            })
```

Comment:
Very-Minor: I would put the increment after the write happened.
```suggestion
            self.device.uartdr.write(|w| unsafe {
                w.data().bits(*c);
                w
            });
            
            bytes_written += 1;
```